controllers/managedocs_controller.go

```diff
@@ -255,3 +323,104 @@ func (r *ManagedOCSReconciler) updateComponents() v1.ComponentStatusMap {
 
 	return components
 }
+
+func (r *ManagedOCSReconciler) isUninstallValid() bool {
+	cmNamespaceName := types.NamespacedName{
+		Name:      r.DeleteConfigMapName,
+		Namespace: r.namespace,
+	}
+
+	configmap := &corev1.ConfigMap{}
+	err := r.Get(r.ctx, cmNamespaceName, configmap)
+	if err == nil {
+		labels := configmap.Labels
+		if _, ok := labels[r.DeleteConfigMapLabelKey]; !ok {
+			return false
+		}
+		return true
+	}
+	return false
+}
+
+func (r *ManagedOCSReconciler) removeOlmComponents() (reconcile.Result, error) {
+
+	r.Log.Info("Deleting subscription")
+	if result, err := r.deleteResource(r.Client, &operators.Subscription{
+		ObjectMeta: metav1.ObjectMeta{Name: r.AddonSubscriptionName, Namespace: r.namespace},
+	}, r.AddonSubscriptionName); err != nil {
+		return ctrl.Result{}, err
+	} else if !result.IsZero() {
+		return result, nil
+	}
+	r.Log.Info("subscription removed successfully")
+
+	r.Log.Info("Deleting CSV")
+	csvList := &operators.ClusterServiceVersionList{}
+	err := r.Client.List(context.Background(), csvList)
+	if err != nil {
+		r.Log.Error(err, "Unable to list CSVs")
+		return ctrl.Result{}, err
+	}
+	for _, csv := range csvList.Items {
+		if strings.HasPrefix(csv.Name, deployerCsvPrefix) {
+			deployerCsvName := csv.Name
+			if result, err := r.deleteResource(r.Client, &operators.ClusterServiceVersion{
+				ObjectMeta: metav1.ObjectMeta{Name: deployerCsvName, Namespace: r.namespace},
+			}, deployerCsvName); err != nil {
```

Comment:
As a matter of readability, let's not create and inline structs inside if conditions. 

```suggestion
			csv := &operators.ClusterServiceVersion{
				ObjectMeta: metav1.ObjectMeta{
					Name: deployerCsvName, 
					Namespace: r.namespace,
				},
			}
			if result, err := r.deleteResource(r.Client, csv, deployerCsvName); err != nil {
```